### PR TITLE
cudf::merge public API now support passing a user stream

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description
Expands the `cudf::merge` function to support a user stream

Found as part of https://github.com/rapidsai/cudf/pull/15982 when building benchmarks

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by mhaseeb123 — APPROVED** [review]

**Comment by mhaseeb123** [comment]
I think the new `STREAM_MERGE_TEST` are failing.

File: cpp/tests/streams/merge_test.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  std::vector<cudf::size_type> key_cols{0};
+  std::vector<cudf::order> column_order;
+  column_order.push_back(cudf::order::ASCENDING);
+  std::vector<cudf::null_order> null_precedence(column_order.size(), cudf::null_order::AFTER);
+
+  cudf::table_view left_view{{leftColWrap1}};
+  cudf::table_view right_view{{rightColWrap1}};
+  cudf::table_view expected{{leftColWrap1}};
+
+  auto result = cudf::merge({left_view, right_view}, key_cols, column_order, null_precedence);
```
Comment (mhaseeb123) [inline cpp/tests/streams/merge_test.cpp:52]: Same change at other places as well.
```suggestion
  auto result = cudf::merge({left_view, right_view}, key_cols, column_order, null_precedence);
  auto result = cudf::merge({left_view, right_view}, key_cols, column_order, null_precedence, cudf::test::get_default_stream());
```

File: cpp/tests/streams/merge_test.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
```
Comment (mhaseeb123) [inline cpp/tests/streams/merge_test.cpp:19]:
```suggestion
#include <cudf_test/column_wrapper.hpp>
#include <cudf_test/default_stream.hpp>
```

File: cpp/tests/streams/merge_test.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  std::vector<cudf::order> column_orders{cudf::order::ASCENDING, cudf::order::DESCENDING};
+  std::vector<cudf::null_order> null_precedences{cudf::null_order::AFTER, cudf::null_order::BEFORE};
+
+  for (auto co : column_orders)
+    for (auto np : null_precedences) {
+      std::vector<cudf::order> column_order{co};
+      std::vector<cudf::null_order> null_precedence{np};
+      auto sorted1 =
+        cudf::sort(cudf::table_view({data1}), column_order, null_precedence)->release();
```
Comment (mhaseeb123) [inline cpp/tests/streams/merge_test.cpp:105]:
```suggestion
        cudf::sort(cudf::table_view({data1}), column_order, null_precedence, cudf::test::get_default_stream())->release();
```

File: cpp/tests/streams/merge_test.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      auto col1 = sorted1.front()->view();
+      auto sorted2 =
+        cudf::sort(cudf::table_view({data2}), column_order, null_precedence)->release();
+      auto col2 = sorted2.front()->view();
+
+      auto result = cudf::merge(
+        {cudf::table_view({col1}), cudf::table_view({col2})}, {0}, column_order, null_precedence);
```
Comment (mhaseeb123) [inline cpp/tests/streams/merge_test.cpp:112]:
```suggestion
        {cudf::table_view({col1}), cudf::table_view({col2})}, {0}, column_order, null_precedence, cudf::test::get_default_stream());
```

File: cpp/tests/streams/merge_test.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      auto sorted_all =
+        cudf::sort(cudf::table_view({all_data->view()}), column_order, null_precedence);
```
Comment (mhaseeb123) [inline cpp/tests/streams/merge_test.cpp:114]:
```suggestion
        cudf::sort(cudf::table_view({all_data->view()}), column_order, null_precedence, cudf::test::get_default_stream());
```

File: cpp/tests/streams/merge_test.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+  for (auto co : column_orders)
+    for (auto np : null_precedences) {
+      std::vector<cudf::order> column_order{co};
+      std::vector<cudf::null_order> null_precedence{np};
+      auto sorted1 =
+        cudf::sort(cudf::table_view({data1}), column_order, null_precedence)->release();
+      auto col1 = sorted1.front()->view();
+      auto sorted2 =
+        cudf::sort(cudf::table_view({data2}), column_order, null_precedence)->release();
```
Comment (mhaseeb123) [inline cpp/tests/streams/merge_test.cpp:108]:
```suggestion
        cudf::sort(cudf::table_view({data2}), column_order, null_precedence, cudf::test::get_default_stream())->release();
```

**Review by mhaseeb123 — CHANGES_REQUESTED** [review]

**Review by mhaseeb123 — COMMENTED** [review]
Thanks for the effort. Waiting for CI tests to pass. Looks good otherwise!

**Review by mhaseeb123 — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Pull request merged**